src/traces/scattermapbox/calc.js

```diff
@@ -0,0 +1,95 @@
+/**
+* Copyright 2012-2016, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+var isNumeric = require('fast-isnumeric');
+
+var Lib = require('../../lib');
+var hasColorscale = require('../../components/colorscale/has_colorscale');
+var makeColorScaleFn = require('../../components/colorscale/make_scale_function');
+var subtypes = require('../scatter/subtypes');
+var calcMarkerColorscale = require('../scatter/marker_colorscale_calc');
+var makeBubbleSizeFn = require('../scatter/make_bubble_size_func');
+
+
+module.exports = function calc(gd, trace) {
+    var len = trace.lon.length,
+        marker = trace.marker;
+
+    var hasMarkers = subtypes.hasMarkers(trace),
+        hasColorArray = (hasMarkers && Array.isArray(marker.color)),
+        hasSizeArray = (hasMarkers && Array.isArray(marker.size)),
+        hasSymbolArray = (hasMarkers && Array.isArray(marker.symbol)),
+        hasTextArray = Array.isArray(trace.text);
+
+    calcMarkerColorscale(trace);
+
+    var colorFn = hasColorscale(trace, 'marker') ?
+            makeColorScaleFn(marker.colorscale, marker.cmin, marker.cmax) :
+            Lib.identity;
+
+    var sizeFn = subtypes.isBubble(trace) ?
+            makeBubbleSizeFn(trace) :
+            Lib.identity;
+
+    var calcTrace = [],
+        cnt = 0;
+
+    // Different than cartesian calc step
+    // as skip over non-numeric lon, lat pairs.
+    // This makes the hover and convert calculations simpler.
```

Comment:
In other words, the `calc` step handles all the per-data-pt computations - unlike for example `gl3d` where per-data-pt computations are done in the `convert` step.

The main reason for utilizing the `calc` more in `mapbox` subplots is too make the cartesian hover code happy - as it looks almost entirely in `gd.calcdata` to determine point position, point colors etc.